storage/storage_image.go

```diff
@@ -90,309 +95,654 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+			}
+			if !has {
+				return errors.Errorf("error determining uncompressed digest for blob %q", blob.Digest.String())
+			}
+			diffID, haveDiffID = s.blobDiffIDs[blob.Digest]
+			if !haveDiffID {
+				return errors.Errorf("we have blob %q, but don't know its uncompressed digest", blob.Digest.String())
+			}
+		}
+		id := diffID.Hex()
+		if lastLayer != "" {
+			id = digest.Canonical.FromBytes([]byte(lastLayer + "+" + diffID.Hex())).Hex()
+		}
+		if layer, err2 := s.imageRef.transport.store.Layer(id); layer != nil && err2 == nil {
+			// There's already a layer that should have the right contents, just reuse it.
+			lastLayer = layer.ID
+			continue
+		}
+		// Check if we cached a file with that blobsum.  If we didn't already have a layer with
+		// the blob's contents, we should have gotten a copy.
+		if filename, ok := s.filenames[blob.Digest]; ok {
+			// Use the file's contents to initialize the layer.
+			file, err2 := os.Open(filename)
+			if err2 != nil {
+				return errors.Wrapf(err2, "error opening file %q", filename)
+			}
+			defer file.Close()
+			diff = file
+		}
+		if diff == nil {
+			// Try to find a layer with contents matching that blobsum.
+			layer := ""
+			layers, err2 := s.imageRef.transport.store.LayersByUncompressedDigest(blob.Digest)
+			if err2 == nil && len(layers) > 0 {
+				layer = layers[0].ID
+			} else {
+				layers, err2 = s.imageRef.transport.store.LayersByCompressedDigest(blob.Digest)
+				if err2 == nil && len(layers) > 0 {
+					layer = layers[0].ID
+				}
+			}
+			if layer == "" {
+				return errors.Wrapf(err2, "error locating layer for blob %q", blob.Digest)
+			}
+			// Use the layer's contents to initialize the new layer.
+			noCompression := archive.Uncompressed
+			diffOptions := &storage.DiffOptions{
+				Compression: &noCompression,
+			}
+			diff, err2 = s.imageRef.transport.store.Diff("", layer, diffOptions)
+			if err2 != nil {
+				return errors.Wrapf(err2, "error reading layer %q for blob %q", layer, blob.Digest)
+			}
+			defer diff.Close()
 		}
+		if diff == nil {
+			// This shouldn't have happened.
+			return errors.Errorf("error applying blob %q: content not found", blob.Digest)
+		}
+		// Build the new layer using the diff, regardless of where it came from.
+		layer, _, err := s.imageRef.transport.store.PutLayer(id, lastLayer, nil, "", false, diff)
+		if err != nil {
+			return errors.Wrapf(err, "error adding layer with blob %q", blob.Digest)
+		}
+		lastLayer = layer.ID
+		addedLayers = append([]string{lastLayer}, addedLayers...)
+	}
+	// If one of those blobs was a configuration blob, then we can try to dig out the date when the image
+	// was originally created, in case we're just copying it.  If not, no harm done.
+	var options *storage.ImageOptions
+	if inspect, err := s.image.Inspect(); err == nil {
+		logrus.Debugf("setting image creation date to %s", inspect.Created)
+		options = &storage.ImageOptions{
+			CreationDate: inspect.Created,
```

Comment:
Ouch ouch ouch. This is the only place which requires constructing a `types.Image` (and actually makes `storageImageUnnamedDestination.GetBlob` necessary for `config.json`).